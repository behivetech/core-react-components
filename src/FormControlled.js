// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {
    debounce,
    difference,
    noop,
    union,
} from 'lodash';

// Components
import Form from './Form';

const DEFAULT_CONTEXT = {
    formDisabled: false,
    getFieldValue: noop,
    getFieldValues: noop,
    initialSubmit: false,
    setFieldError: noop,
    setFieldValue: noop,
    setFieldValueDebounced: noop,
    unsetFieldError: noop,
};

export const FormContext = React.createContext(DEFAULT_CONTEXT);

export default class FormControlled extends Component {
    constructor(props) {
        super(props);

        this.state = {
            formDisabled: this.props.disabled,
            initialSubmit: false,
            fieldErrors: [],
            fieldValues: {},
            submitting: false,
        };
        this.setFieldValueDebounced = debounce(this.setFieldValue, 300, {trailing: true});
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    UNSAFE_componentWillMount() {
        const {fieldValues} = this.props;

        if (fieldValues) {
            this.setState({fieldValues});
        }
    }

    getFieldValue(fieldName) {
        return this.state.fieldValues[fieldName];
    }

    getFieldValues() {
        return this.state.fieldValues;
    }

    setFieldError(fieldName) {
        this.setState({fieldErrors: union(this.state.fieldErrors, [fieldName])});
    }

    unsetFieldError(fieldName) {
        this.setState({fieldErrors: difference(this.state.fieldErrors, [fieldName])});
    }

    setFieldValue(fieldName, value) {
        if (fieldName && value !== this.getFieldValue(fieldName)) {
            this.setState({
                fieldValues: {
                    ...this.state.fieldValues,
                    [fieldName]: value,
                },
            });
        }
    }

    hasErrors() {
        return (!!this.state.fieldErrors.length || !!this.props.formError);
    }

    handleSubmit(event) {
        this.setState({initialSubmit: true});

        if (!this.hasErrors()) {
            this.setState({
                formDisabled: true,
                submitting: true,
            });

            this.props.onSubmit(this.state.fieldValues, this.handleSubmitted);
        }
    }

    handleSubmitted() {
        this.setState({
            formDisabled: false,
            submitting: false,
        });
    }


    render() {
        const {children, className, formError, id, name} = this.props;
        const {formDisabled, initialSubmit} = this.state;
        const formContext = {
            formDisabled,
            initialSubmit,
            getFieldValue: this.getFieldValue.bind(this),
            getFieldValues: this.getFieldValues.bind(this),
            setFieldError: this.setFieldError.bind(this),
            setFieldValue: this.setFieldValue.bind(this),
            setFieldValueDebounced: this.setFieldValueDebounced.bind(this),
            unsetFieldError: this.unsetFieldError.bind(this),
        };

        return (
            <FormContext.Provider value={formContext}>
                <Form
                    className={classnames('form-controlled', className)}
                    id={id}
                    name={name}
                    onSubmit={this.handleSubmit}
                >
                    {children}
                    {(formError) ? <div>{formError}</div> : null}
                </Form>
            </FormContext.Provider>
        );
    }
}

FormControlled.propTypes = {
    /** The children to go in the form */
    children: PropTypes.node,
    /** Used to add attitional styling from parent component */
    className: PropTypes.string,
    /** Disable all elements of the form */
    disabled: PropTypes.bool,
    /** Any form type errors to help with it submitting and showing errors */
    formError: PropTypes.string,
    /**
        Values to populate the fields going by the same name you use
        as the name prop for the field and using the FormField component
    */
    fieldValues: PropTypes.objectOf(PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.number,
        PropTypes.string,
    ])),
    /** id attribute for the form */
    id: PropTypes.string,
    /**
        Standard name you would give a form using html. Helps with react refs.
        It's ideal to keep these names unique per what's displayed on the current
        page because of the potential refs created.
        (refs are coming from a FormRefProvider component to have the ability to
        submit the form outside of the form).
    */
    // TODO: need to figure out the ref thing and disabling components that would
    // potentially submit this form from outside of the form. Probably pass the
    // values down through the value argument in the provider and callback functions.
    name: PropTypes.string.isRequired,
    /**
        Functionality to run when this Form component meets all requirements
        such as passsing validation tests.
    */
    onSubmit: PropTypes.func,
};

FormControlled.defaultProps = {
    disabled: false,
    onSubmit: noop,
};
