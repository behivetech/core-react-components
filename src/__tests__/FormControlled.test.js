import FormControlled from '../FormControlled';

describe('FormControlled', () => {
    const getProps = (updateProps) => {
        return {
            className: 'mock-class',
            name: 'mockForm',
            ...updateProps,
        };
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with the correct state and props', () => {
        const mockProps = {fieldValues: {mockField: 'mock value'}};
        const wrapper = mount(<FormControlled {...getProps(mockProps)} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.state()).toMatchSnapshot();
    });

    it('should set and unset a field error', () => {
        const wrapper = shallow(<FormControlled {...getProps()} />);
        const wrapperInstance = wrapper.instance();

        wrapperInstance.setFieldError('mockField1');
        expect(wrapperInstance.state.fieldErrors[0]).toBe('mockField1');
        wrapperInstance.unsetFieldError('mockField1');
        expect(wrapperInstance.state.fieldErrors.length).toBe(0);
    });

    it('should show if there are errors or not', () => {
        const wrapper = shallow(<FormControlled {...getProps()} />);
        const wrapperInstance = wrapper.instance();

        expect(wrapperInstance.hasErrors()).toBe(false);
        wrapperInstance.setFieldError('mockField1');
        expect(wrapperInstance.hasErrors()).toBe(true);
        wrapperInstance.setFieldError('mockField2');
        expect(wrapperInstance.hasErrors()).toBe(true);
        wrapperInstance.unsetFieldError('mockField1');
        expect(wrapperInstance.hasErrors()).toBe(true);
        wrapperInstance.unsetFieldError('mockField2');
        expect(wrapperInstance.hasErrors()).toBe(false);
        wrapper.setProps({formError: 'mock error'});
        expect(wrapperInstance.hasErrors()).toBe(true);
    });

    it('should or should not set a field value when conditions are met', () => {
        const mockProps = getProps({fieldValues: {mockField: 'mock value'}});
        const wrapper = shallow(<FormControlled {...mockProps} />);
        const wrapperInstance = wrapper.instance();
        const mockSetState = jest.fn();

        wrapperInstance.setState = mockSetState;
        wrapperInstance.setFieldValue('mockField', 'mock value');
        expect(mockSetState).not.toHaveBeenCalled();
        wrapperInstance.setFieldValue('mockField', 'mock value 1');
        expect(mockSetState).toHaveBeenCalledWith(
            {fieldValues: {mockField: 'mock value 1'}}
        );
    });

    it('should handle a submit correctly', () => {
        const mockEvent = {preventDefault: jest.fn()};
        const mockProps = getProps({onSubmit: jest.fn()});
        const wrapper = shallow(<FormControlled {...mockProps} />);
        const wrapperInstance = wrapper.instance();
        const initialState = wrapperInstance.state;

        wrapperInstance.handleSubmit(mockEvent);
        expect(wrapperInstance.state.formDisabled).toBe(true);
        expect(wrapperInstance.state.initialSubmit).toBe(true);
        expect(wrapperInstance.state.submitting).toBe(true);
        expect(mockProps.onSubmit).toBeCalled();

        wrapper.setState(initialState);
        jest.clearAllMocks();
        wrapperInstance.setFieldError('mockField');
        wrapperInstance.handleSubmit(mockEvent);
        expect(wrapperInstance.state.formDisabled).toBe(false);
        expect(wrapperInstance.state.initialSubmit).toBe(true);
        expect(wrapperInstance.state.submitting).toBe(false);
        expect(mockProps.onSubmit).not.toHaveBeenCalled();

        wrapper.setState(initialState);
        jest.clearAllMocks();
        wrapper.setProps({formError: 'mock form error'});
        wrapperInstance.handleSubmit(mockEvent);
        expect(wrapperInstance.state.formDisabled).toBe(false);
        expect(wrapperInstance.state.initialSubmit).toBe(true);
        expect(wrapperInstance.state.submitting).toBe(false);
        expect(mockProps.onSubmit).not.toHaveBeenCalled();
    });

    it('should return the right values from the context state', () => {
        const mockStateValues = {
            fieldValues: {mockField: 'mock value'},
        };
        const wrapper = shallow(<FormControlled {...getProps()} />);
        const wrapperInstance = wrapper.instance();

        wrapper.setState(mockStateValues);
        expect(wrapperInstance.getFieldValue('mockField')).toBe('mock value');
        expect(wrapperInstance.getFieldValues()).toBe(mockStateValues.fieldValues);
        expect(wrapperInstance.state.initialSubmit).toBe(false);
    });
});
