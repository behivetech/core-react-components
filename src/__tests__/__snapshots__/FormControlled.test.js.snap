// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControlled should render with the correct state and props 1`] = `
<FormControlled
  className="mock-class"
  disabled={false}
  fieldValues={
    Object {
      "mockField": "mock value",
    }
  }
  name="mockForm"
  onSubmit={[Function]}
>
  <Form
    className="form-controlled mock-class"
    name="mockForm"
    onSubmit={[Function]}
  >
    <form
      className="form form-controlled mock-class"
      name="mockForm"
      onSubmit={[Function]}
    />
  </Form>
</FormControlled>
`;

exports[`FormControlled should render with the correct state and props 2`] = `
Object {
  "fieldErrors": Array [],
  "fieldValues": Object {
    "mockField": "mock value",
  },
  "formDisabled": false,
  "initialSubmit": false,
  "submitting": false,
}
`;
